tools/splitblocks/main.go

```diff
@@ -0,0 +1,232 @@
+		return fmt.Errorf("max-block-duration should be aligned to hours")
+	}
+	if 24*time.Hour%c.maxBlockDuration.Truncate(time.Hour) != 0 {
+		return fmt.Errorf("max-block-duration should divide 24h without remainder")
+	}
+	if c.outputDir == "" {
+		return fmt.Errorf("output-dir is required")
+	}
+	if c.blockConcurrency < 1 {
+		return fmt.Errorf("block-concurrency must be positive")
+	}
+	return nil
+}
+
+func main() {
+	// Clean up all flags registered via init() methods of 3rd-party libraries.
+	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	cfg := config{}
+	cfg.registerFlags(flag.CommandLine)
+
+	// Parse CLI arguments.
+	if err := flagext.ParseFlagsWithoutArguments(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	if err := cfg.validate(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	logger := log.NewLogfmtLogger(os.Stdout)
+	logger = log.With(logger, "ts", log.DefaultTimestampUTC)
+
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	if err := splitBlocks(ctx, cfg, logger); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+func splitBlocks(ctx context.Context, cfg config, logger log.Logger) error {
+	bkt, err := bucket.NewClient(ctx, cfg.bucket, "bucket", logger, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to create bucket")
+	}
+
+	// Helpful when not specifying -backend=filesystem (since it can use -filesystem.dir)
+	if cfg.bucketPrefix != "" {
+		bkt = bucket.NewPrefixedBucketClient(bkt, cfg.bucketPrefix)
+	}
+
+	var blocks []string
+	if len(cfg.blocks) > 0 {
+		blocks = cfg.blocks
+	} else {
+		blocks, err = listBlocks(ctx, bkt)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to list blocks", "err", err)
+			return errors.Wrapf(err, "failed to list blocks")
+		}
+	}
+
+	return concurrency.ForEachUser(ctx, blocks, cfg.blockConcurrency, func(ctx context.Context, blockString string) error {
+		blockID, err := ulid.Parse(blockString)
+		if err != nil {
+			return err
+		}
+
+		logger := log.With(logger, "block", blockString)
+		blockMeta, err := block.DownloadMeta(ctx, logger, bkt, blockID)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to read block's meta.json file", "err", err)
+			return err
+		}
+
+		blockMinTime := timestamp.Time(blockMeta.MinTime)
+		blockMaxTime := timestamp.Time(blockMeta.MaxTime)
+		level.Info(logger).Log("block_min_time", blockMinTime, "block_max_time", blockMaxTime)
+
+		if blockMinTime.After(blockMaxTime) {
+			level.Error(logger).Log("msg", "block has an invalid minTime greater than maxTime")
+			return fmt.Errorf("block has an invalid minTime greater than maxTime")
+		}
+
+		allowedMaxTime := blockMinTime.Truncate(cfg.maxBlockDuration).Add(cfg.maxBlockDuration)
+		if !blockMaxTime.After(allowedMaxTime) {
+			level.Info(logger).Log("msg", "block does not need to be split")
```

Comment:
Shall we download such blocks to output directory anyway? That way after running the command, only blocks from output directory can be backfilled.

Reply:
I added a new flag to do this, `--full`.